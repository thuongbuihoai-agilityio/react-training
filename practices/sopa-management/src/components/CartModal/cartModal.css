.cart {
  width: var(--w-xs-large);
  padding: var(--py-extra-normal) var(--px-extra-normal) 0;
  position: absolute;
  z-index: 10;
  right: 0;
  top: 0;
  bottom: 0;
  background-color: var(--light);
}

.cart-header,
.cart-item,
.cart-footer {
  display: flex;
  justify-content: space-between;
}

.cart-image {
  margin-right: var(--m-medium);
  min-width: var(--w-extra-small);
  height: var(--h-extra-normal);
  border-radius: var(--rounded-secondary);
  border: 1px solid var(--border-primary);
  background-color: var(--bg-primary);
}

.cart-image-item {
  object-fit: contain;
  width: var(--w-extra-small);
  height: var(--h-extra-normal);
}

.card-body {
  margin-top: var(--m-7xl);
  height: 78%;
}

.cart-item {
  margin-bottom: var(--m-3xl);
}

.cart-info {
  display: flex;
}

.card-quantity {
  z-index: 5;
  display: flex;
}

.cart-color {
  margin: var(--m-extra-regular) 0 var(--m-regular);
  font-size: var(--text-regular);
}

.cart-size {
  margin-bottom: var(--m-extra-medium);
  font-size: var(--text-regular);
}

.cart-btn {
  margin-left: var(--m-extra-regular);
}

.cart-footer {
  margin-top: var(--m-extra-medium);
}

.cart-text {
  font-weight: var(--font-semi-bold);
  font-size: var(--text-base);
  line-height: var(--leading-primary);
}

.cart-message {
  color: var(--border-tertiary);
}

.overlay {
  position: fixed;
  z-index: 5;
  top:0;
  left:0;
  right:0;
  bottom:0;
  background-color: var(--bg-secondary);
}

.cart-update,
.cart-disable {
  position: absolute;
  bottom: 25px;
  right: 35px;
  width: 612px;
  padding: var(--p-extra-small) 0;
  border-radius: var(--rounded-secondary);
  color: var(--light);
  border: 1px solid var(--text-quaternary);
  background: var(--dark);
}

.cart-disable {
  cursor: not-allowed;
}
