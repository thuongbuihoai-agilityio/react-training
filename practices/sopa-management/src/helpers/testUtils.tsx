import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import {
  QueryClient,
  QueryClientProvider
} from 'react-query';
import {
  BrowserRouter,
  Router
} from 'react-router-dom';

export const renderRouterTest = (component: JSX.Element) => {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {component}
    </Router>
  );
};

export const renderWithRouterAndQuery = (children: JSX.Element) => {
  const queryClient = new QueryClient();

  return render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </BrowserRouter>
  );
};
