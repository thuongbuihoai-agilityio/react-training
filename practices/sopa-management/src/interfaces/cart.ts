export const enum QuantityType {
  increment = 'increment',
  decrement = 'decrement'
}
