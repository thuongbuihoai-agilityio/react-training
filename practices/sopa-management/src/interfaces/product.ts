export interface Product {
  id: string;
  name: string;
  price: number;
  color: string;
  image: {
    url: string;
    alt: string;
  };
  size: string;
  description?: string;
  quantity: number;
};
