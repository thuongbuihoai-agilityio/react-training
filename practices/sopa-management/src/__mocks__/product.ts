import { Product } from "../interfaces/product";

const MOCK_PRODUCTS: Product[] = [
  {
    name: 'Black shoes',
    price: 87,
    size: 'Small',
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Black',
    image: {
      url: '/images/products/black-shoes.png',
      alt: 'This is black shoes'
    },
    id: '1',
    quantity: 2
  },
  {
    name: 'Blue shoes',
    price: 45,
    size: 'Medium',
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Blue',
    image: {
      url: '/images/products/blue-shoes.png',
      alt: 'This is blue shoes'
    },
    id: '2',
    quantity: 1
  }
];

const MOCK_PRODUCT: Product = {
  name: 'Black shoes',
  price: 87,
  size: 'Large',
  description: 'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
  color: 'Black',
  image: {
    url: '/images/products/black-shoes.png',
    alt: 'This is black shoes'
  },
  id: '1',
  quantity: 3
};

export {
  MOCK_PRODUCTS,
  MOCK_PRODUCT
};
