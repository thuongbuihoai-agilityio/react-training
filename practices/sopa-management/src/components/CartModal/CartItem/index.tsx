import {
  useCallback,
  useState
} from 'react';
import toast from 'react-hot-toast';
import { shallow } from 'zustand/shallow';

// Interfaces
import { Product } from '@interfaces/product';
import { QuantityType } from '@interfaces/cart';

// Components
import {
  Button,
  ButtonType,
  Counter,
  Icon,
  IconType,
  Image,
  PopupDelete,
  Price,
  PriceType,
  Text,
  ThemeType
} from '@components/common';

// Stores
import { useCartStore } from '@stores/cart';
import { CONFIRM_MESSAGE } from '@constants/validate';

// Hooks
import { useMutationDeleteProduct } from '@hooks/useMutate';

interface CartItemProps {
  cartItem: Product;
  onChangeQuantity?: () => void;
}

const CartItem = ({
  cartItem,
  onChangeQuantity = () => {}
}: CartItemProps) => {
  const [openModalConfirm, setOpenModalConfirm] = useState<boolean>(false);
  const { mutate: deleteProduct } = useMutationDeleteProduct();

  const [updateQuantity, deleteProductInCart] = useCartStore(
    (state) => [state.updateQuantity, state.deleteProductInCart],
    shallow
  );

  const handleIncrement = useCallback(() => {
    updateQuantity(cartItem.id, QuantityType.increment);
    onChangeQuantity();
  }, [cartItem]);

  const handleDecrement = useCallback(() => {
    updateQuantity(cartItem.id, QuantityType.decrement);
    onChangeQuantity();
  }, [cartItem]);

  const handleOpenModalConfirm = useCallback(() => {
    setOpenModalConfirm(!openModalConfirm);
  }, [openModalConfirm]);

  const handleDeleteCart = useCallback(() => {
    deleteProduct(cartItem.id, {
      onError: (error) => {
        toast.error((error as { message: string }).message);
      },
      onSuccess: () => {
        deleteProductInCart(cartItem.id);
        toast.success(CONFIRM_MESSAGE.DELETE_SUCCESS);
      }
    });
  }, []);

  return (
    <div data-testid='cart-item' className='cart-item'>
      <div className='cart-info'>
        <div className='cart-image'>
          <Image
            className='cart-image-item'
            url={cartItem?.image?.url}
            alt={cartItem.image.alt}
            width={130}
            height={130}
          />
        </div>
        <div className='cart-description'>
          <Text text={cartItem.name} className='cart-text' />
          <Text
            text={`Color: ${cartItem.color}`}
            type={ThemeType.highlightPrimary}
            className='cart-color'
          />
          <Text
            text={`Size: ${cartItem.size}`}
            type={ThemeType.highlightPrimary}
            className='cart-size'
          />
          <div className='card-quantity'>
            <Counter
              value={cartItem?.quantity}
              onDecrement={handleDecrement}
              onIncrement={handleIncrement}
            />
            <Button
              ariaLabel='Trash'
              children={<Icon iconName={IconType.trash} />}
              type={ButtonType.btnIconSecondary}
              className='cart-btn'
              onClick={handleOpenModalConfirm}
            />
            {openModalConfirm && (
              <PopupDelete
                title={CONFIRM_MESSAGE.CONFIRM_DELETE}
                onCancel={handleOpenModalConfirm}
                onDelete={handleDeleteCart}
              />
            )}
          </div>
        </div>
      </div>
      <Price
        value={cartItem.price}
        type={PriceType.tertiary}
        className='cart-price'
      />
    </div>
  );
};

export default CartItem;
