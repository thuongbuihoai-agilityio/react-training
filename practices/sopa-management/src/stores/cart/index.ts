import { create } from 'zustand';

// Interfaces
import { Product } from '@interfaces/product';
import { QuantityType } from '@interfaces/cart';

type CartType = {
  cart: Product[];
  setCart: (value: Product[]) => void;
  addToCart: (product: Product, size: string) => void;
  updateQuantity: (id: string, quantityType: QuantityType) => void;
  deleteProductInCart: (id: string) => void;
};

export const useCartStore = create<CartType>()((set) => ({
  cart: [],
  setCart: (data: Product[]) => set(() => ({ cart: data })),
  addToCart: (product: Product, size: string) => {
    set((state) => {
      const currentCart = state.cart || [];
      const existingProduct = currentCart.find(
        (item: Product) => item.id === product.id && item.size === size
      );

      let updatedCart;
      if (existingProduct) {
        updatedCart = currentCart.map((item: Product) =>
          item.id === product.id && item.size === size
            ? { ...item, quantity: item.quantity + 1 }
            : item
        );
      } else {
        updatedCart = [...currentCart, { ...product, quantity: 1, size: size }];
      }

      set({ cart: updatedCart });
      return { cart: updatedCart };
    });
  },

  updateQuantity: (id: string, quantityType: QuantityType) => {
    set((state) => {
      const currentCart = state.cart || [];
      const updatedCart = currentCart.map((item: Product) => {
        const newQuantity =
          quantityType === QuantityType.increment
            ? item.quantity + 1
            : item.quantity - 1;
        return item.id === id ? { ...item, quantity: newQuantity } : item;
      });

      set({ cart: updatedCart });
      return { cart: updatedCart };
    });
  },

  deleteProductInCart: (id: string) =>
    set((state) => ({
      cart: state.cart.filter((product) => product.id !== id)
    }))
}));
