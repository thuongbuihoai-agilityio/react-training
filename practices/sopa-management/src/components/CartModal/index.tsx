import { memo, useCallback, useState } from 'react';
import toast from 'react-hot-toast';

// Components
import {
  Button,
  ButtonType,
  Icon,
  IconType,
  Price,
  PriceType,
  Text
} from '@components/common';
import CartItem from './CartItem';

// Interfaces
import { Product } from '@interfaces/product';

// Helpers
import { totalPrices } from '@helpers/common';

// Constants
import { CONFIRM_MESSAGE } from '@constants/validate';


// Hooks
import { useMutationEditProductInCart } from '@hooks/useMutate';

// Styles
import './cartModal.css';
import { SizeType } from '@components/common/Text';

interface CartModalProps {
  carts: Product[];
  onToggleModal?: () => void;
}
const CartModal = ({
  carts,
  onToggleModal
}: CartModalProps) => {
  const { mutate: putProduct } = useMutationEditProductInCart();
  const [checkChangesQuantity, setCheckChangesQuantity] = useState(false);

  const handleCheckChangeQuantity = useCallback(() => {
    setCheckChangesQuantity(true)
  }, [checkChangesQuantity])

  const handleUpdateProduct = async (products: Product[]) => {
    const updatedProducts = [];
    for (const product of products) {
      const updatedProduct = await putProduct({
        ...product,
        quantity: product.quantity
      },
      {
        onError: (error) => {
          toast.error((error as { message: string }).message);
        },
        onSuccess: () => {
          updatedProducts.push(updatedProduct);
          toast.success(CONFIRM_MESSAGE.UPDATE_SUCCESS);
        },
      });
    }
  };

  return (
    <div data-testid='cart-modal' className='overlay'>
      <div className='cart'>
        <div className='cart-header'>
          <Text text='Cart' className='cart-text' />
          <Button
            ariaLabel='Close'
            children={<Icon iconName={IconType.close} />}
            type={ButtonType.btnIconPrimary}
            onClick={onToggleModal}
          />
        </div>
        <div className='card-body'>
          {carts.length
            ? <>
                {carts?.map((cartItem: Product) => (
                  <CartItem
                    key={cartItem.id}
                    cartItem={cartItem}
                    onChangeQuantity={handleCheckChangeQuantity}
                  />
                ))}
              </>
            : <Text
                text='No products in cart'
                type={SizeType.extraRegular}
                className='cart-message'
              />
            }
        </div>
        <hr />
        <div className='cart-footer'>
          <Text text='Subtotal' className='cart-text' />
          <Price value={totalPrices(carts)} type={PriceType.tertiary} />
        </div>
        <Button
          children='Update'
          className={checkChangesQuantity ? 'cart-update' : 'cart-disable'}
          disable={!checkChangesQuantity}
          onClick={() => handleUpdateProduct(carts)}
        />
      </div>
    </div>
  );
};

export default memo(CartModal);
