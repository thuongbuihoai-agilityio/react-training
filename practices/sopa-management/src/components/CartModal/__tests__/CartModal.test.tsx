// Libs
import '@testing-library/jest-dom';

// Helpers
import { renderWithRouterAndQuery } from '@helpers/testUtils';

// Components
import CartModal from '@components/CartModal';
import { MOCK_PRODUCTS } from '@mocks/product';

// Stores
import { useCartStore } from '@stores/cart';

describe('Modal component', () => {
  beforeEach(() => {
    useCartStore.setState({
      cart: MOCK_PRODUCTS,
    });
  });

  test('Should render Modal component by default', () => {
    const { getByTestId } = renderWithRouterAndQuery(<CartModal carts={MOCK_PRODUCTS} />);
    const cartModal = getByTestId('cart-modal');
    expect(cartModal).toBeInTheDocument();
  });

  test('should render CartModal with cart items when carts is true', () => {
    const { getByTestId, getByText } = renderWithRouterAndQuery(<CartModal carts={MOCK_PRODUCTS} />);
    const cartModal = getByTestId('cart-modal');
    expect(cartModal).toBeInTheDocument();
    expect(getByText('Cart')).toBeInTheDocument();
  });

  test('Matches snapshot', () => {
    const { container } = renderWithRouterAndQuery(<CartModal carts={MOCK_PRODUCTS} />);
    expect(container).toMatchSnapshot();
  });
});
