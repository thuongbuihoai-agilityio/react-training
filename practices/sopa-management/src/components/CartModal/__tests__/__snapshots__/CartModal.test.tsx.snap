// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component Matches snapshot 1`] = `
<div>
  <div
    class="overlay"
    data-testid="cart-modal"
  >
    <div
      class="cart"
    >
      <div
        class="cart-header"
      >
        <p
          class="cart-text text-"
          data-testid="text"
        >
          Cart
        </p>
        <button
          aria-label="Close"
          class=" btn btn-icon-primary"
          data-testid="button"
        >
          <div
            class="undefined icon-close"
            data-testid="icon"
          />
        </button>
      </div>
      <div
        class="card-body"
      >
        <div
          class="cart-item"
          data-testid="cart-item"
        >
          <div
            class="cart-info"
          >
            <div
              class="cart-image"
            >
              <img
                alt="This is black shoes"
                class="cart-image-item"
                data-testid="image"
                src="/images/products/black-shoes.png"
              />
            </div>
            <div
              class="cart-description"
            >
              <p
                class="cart-text text-"
                data-testid="text"
              >
                Black shoes
              </p>
              <p
                class="cart-color text-highlight-primary"
                data-testid="text"
              >
                Color: Black
              </p>
              <p
                class="cart-size text-highlight-primary"
                data-testid="text"
              >
                Size: Small
              </p>
              <div
                class="card-quantity"
              >
                <div
                  class="counter"
                  data-testid="counter"
                >
                  <button
                    aria-label="Decrement"
                    class="counter-minus btn btn-icon-primary"
                    data-testid="button"
                  />
                  <input
                    aria-hidden="true"
                    class=" counter-input input-"
                    data-testid="input"
                    id=""
                    name=""
                    type="number"
                    value="2"
                  />
                  <button
                    aria-label="Increment"
                    class="counter-plus btn btn-icon-primary"
                    data-testid="button"
                  />
                </div>
                <button
                  aria-label="Trash"
                  class="cart-btn btn btn-icon-secondary"
                  data-testid="button"
                >
                  <div
                    class="undefined icon-trash"
                    data-testid="icon"
                  />
                </button>
              </div>
            </div>
          </div>
          <p
            class="cart-price price-tertiary"
            data-testid="price"
          >
            $
            87
          </p>
        </div>
        <div
          class="cart-item"
          data-testid="cart-item"
        >
          <div
            class="cart-info"
          >
            <div
              class="cart-image"
            >
              <img
                alt="This is blue shoes"
                class="cart-image-item"
                data-testid="image"
                src="/images/products/blue-shoes.png"
              />
            </div>
            <div
              class="cart-description"
            >
              <p
                class="cart-text text-"
                data-testid="text"
              >
                Blue shoes
              </p>
              <p
                class="cart-color text-highlight-primary"
                data-testid="text"
              >
                Color: Blue
              </p>
              <p
                class="cart-size text-highlight-primary"
                data-testid="text"
              >
                Size: Medium
              </p>
              <div
                class="card-quantity"
              >
                <div
                  class="counter"
                  data-testid="counter"
                >
                  <button
                    aria-label="Decrement"
                    class="counter-minus-disable btn btn-icon-primary"
                    data-testid="button"
                    disabled=""
                  />
                  <input
                    aria-hidden="true"
                    class=" counter-input input-"
                    data-testid="input"
                    id=""
                    name=""
                    type="number"
                    value="1"
                  />
                  <button
                    aria-label="Increment"
                    class="counter-plus btn btn-icon-primary"
                    data-testid="button"
                  />
                </div>
                <button
                  aria-label="Trash"
                  class="cart-btn btn btn-icon-secondary"
                  data-testid="button"
                >
                  <div
                    class="undefined icon-trash"
                    data-testid="icon"
                  />
                </button>
              </div>
            </div>
          </div>
          <p
            class="cart-price price-tertiary"
            data-testid="price"
          >
            $
            45
          </p>
        </div>
      </div>
      <hr />
      <div
        class="cart-footer"
      >
        <p
          class="cart-text text-"
          data-testid="text"
        >
          Subtotal
        </p>
        <p
          class=" price-tertiary"
          data-testid="price"
        >
          $
          219
        </p>
      </div>
      <button
        aria-label=""
        class="cart-disable btn btn-"
        data-testid="button"
        disabled=""
      >
        Update
      </button>
    </div>
  </div>
</div>
`;
